Fix issues related to GPUNetIO

foraxe opened the pull request:
## What?
This PR fixs the GPUNetIO backend to make nixlbench run on intranode.

## Why?
The DOCA couldn't start transfer data in our server envs, with following log:
```shell
prepare_write radd 7f8637fff010 rkey 130100 ladd 7f6180000000 lkey d2100100 size 4096     
prepare_write radd 7f8637fff010 rkey 130100 ladd 7f6180000000 lkey d2100100 size 4096
>>>>>>> CUDA rdma write kernel pos 0 posted 512 buffers from base_wqe_idx 0               
got completion with err: syndrome=0x5, vendor_err_synd=0xf9, hw_err_synd=0, hw_synd_type=0, wqe_counter=65281 wqe_qpn=ea030008                                                      
kernel_progress: block 0 error CQE! poll_status -5 wqe 511 index 0     
```
I had tried many versions of code in community, but ended up with above log. 
So, i am uploading an working version which ran on our local servers. 
Ref:
https://github.com/ai-dynamo/nixl/issues/952#issuecomment-3489060163
https://github.com/ai-dynamo/nixl/issues/788
https://github.com/NVIDIA-DOCA/gpunetio/issues/2#issuecomment-3641431377

## How to run: 
`NIXL_DOCASIM_IPV4_OVERRIDE` use the inet ip of mlx5_bond0. One can obtain it by running command `ip a`
* Run on node 0:
```shell
ETCDCTL_API=3 etcdctl del "" --from-key=true 
export NIXL_GPUNETIO_DEBUG_DUMP=1
    export NIXL_GPUNETIO_SWAP_KEYS=1
    export NIXL_LOG_LEVEL=DEBUG
    export NIXL_DOCASIM_IPV4_OVERRIDE=200.x.7.238
cd /workspace/nixl/benchmark/nixlbench/build
LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/opt/gdrcopy/lib:/opt/mellanox/doca:/opt/mellanox/doca/lib/x86_64-linux-gnu NIXL_PLUGIN_DIR=/usr/local/nixl/lib/x86_64-linux-gnu/plugins CUDA_MODULE_LOADING=EAGER ./nixlbench --etcd-endpoints http://10.x.x.21:2379 --backend=GPUNETIO --initiator_seg_type=VRAM --target_seg_type=DRAM --runtime_type=ETCD --gpunetio_device_list=0 --device_list=mlx5_bond_0 --start_batch_size=1 --max_batch_size=1 --total_buffer_size=131072 --max_block_size 16384 --op_type=READ
```
* Run for intranode: node 0, use above codes. 
* Result:
<img width="2238" height="620" alt="image" src="https://github.com/user-attachments/assets/fa413af2-3a69-4539-807e-9eb661c7b9f4" />


* If you want to test run for internode: node 1 , change NIXL_DOCASIM_IPV4_OVERRIDE: 
```shell
    export NIXL_GPUNETIO_DEBUG_DUMP=1
    export NIXL_GPUNETIO_SWAP_KEYS=1
export NIXL_LOG_LEVEL=DEBUG
    export NIXL_DOCASIM_IPV4_OVERRIDE=200.x.7.238 #change this addr to try on node1
cd /workspace/nixl/benchmark/nixlbench/build
LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/opt/gdrcopy/lib:/opt/mellanox/doca:/opt/mellanox/doca/lib/x86_64-linux-gnu NIXL_PLUGIN_DIR=/usr/local/nixl/lib/x86_64-linux-gnu/plugins CUDA_MODULE_LOADING=EAGER ./nixlbench --etcd-endpoints http://10.x.x.21:2379 --backend=GPUNETIO --initiator_seg_type=VRAM --target_seg_type=DRAM --runtime_type=ETCD --gpunetio_device_list=0 --device_list=mlx5_bond_0 --start_batch_size=1 --max_batch_size=1 --total_buffer_size=131072 --max_block_size 16384 --op_type=READ
```

cc @e-ago @ovidiusm

foraxe committed:
Fix issues related to GPUNetIO

brminich commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -1153,13 +1241,36 @@ nixlDocaEngine::prepXfer(const nixl_xfer_op_t &operation,
-        }
+
+            //uint32_t lkey = xferReqRingCpu[pos].lkey[idx];
+            //uint32_t rkey = xferReqRingCpu[pos].rkey[idx];
```
remove

brminich commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -1153,13 +1241,36 @@ nixlDocaEngine::prepXfer(const nixl_xfer_op_t &operation,
+                NIXL_INFO << "[dbg] remote_desc_addr=" << format_hex(dbg_rbuf_desc)
+                          << " remote_mr_addr=" << format_hex(dbg_rbuf_mr)
+                          << " used=mr_addr";
+            }
```
imo it should be NIXL_TRACE

brminich commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -448,7 +487,11 @@ nixlDocaEngine::nixlDocaInitNotif(const std::string &remote_agent, doca_dev *dev
     notifMap[remote_agent] = notif;
     ((volatile struct docaNotif *)notif_fill_cpu)->msg_buf = (uintptr_t)notif->recv_addr;
-    ((volatile struct docaNotif *)notif_fill_cpu)->msg_lkey = notif->recv_mr->get_lkey();
+//    ((volatile struct docaNotif *)notif_fill_cpu)->msg_lkey = notif->recv_mr->get_lkey();
```
```suggestion

```

brminich commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -47,6 +58,24 @@ nixlDocaEngine::nixlDocaEngine(const nixlBackendInitParams *init_params)
+        else
+            swap_keys_config = true;
+    }
+    const char *env_dbg = std::getenv("NIXL_GPUNETIO_DEBUG_DUMP");
```
why not using NIXL_DEBUG instead?

foraxe committed:
Implemented all 13 requested PR comment fixes by removing leftover debug/dead code and switching non-error output to logging macros.

- Replaced ad-hoc prints with NIXL_DEBUG/NIXL_TRACE in src/plugins/gpunetio/gpunetio_backend.cpp:74, src/plugins/gpunetio/gpunetio_backend.cpp:976, src/plugins/gpunetio/gpunetio_backend.cpp:1262, and src/plugins/gpunetio/verbs/verbs.cpp:469.
- Removed the commented-out lines the reviewer asked to delete (e.g., msg_lkey/lkey_notif/old parsing block), and cleaned up the “dead code” block in loadRemoteMD.
- Applied the std::memcpy suggestion and fixed indentation in src/plugins/gpunetio/gpunetio_backend.cpp (around progressThreadStart).

foraxe commented on src/plugins/gpunetio/verbs/verbs.cpp:
```diff
@@ -463,8 +463,23 @@ mr::mr(doca_gpu *gpu_dev_, void *addr_, uint32_t elem_num_, size_t elem_size_, s
+            << " lkey 0x" << std::hex << (uint32_t)lkey << std::dec
+            << " rkey 0x" << std::hex << (uint32_t)rkey << std::dec;
+        NIXL_INFO << oss.str();
+    }
```
Switched the debug-dump MR log from NIXL_INFO to NIXL_DEBUG (only emits when debug logging is enabled).

foraxe replied:
the env_dbg is currently kept for seperating debugg outputs.

brminich replied:
why do you want to separate? Once gpunetio plugin is selected it is not supposed to interfere with others.

foraxe commented on src/plugins/gpunetio/verbs/verbs.cpp:
```diff
@@ -463,8 +463,23 @@ mr::mr(doca_gpu *gpu_dev_, void *addr_, uint32_t elem_num_, size_t elem_size_, s
-    lkey = htobe32(ibmr->lkey);
-    rkey = htobe32(ibmr->rkey);
+//    lkey = htobe32(ibmr->lkey);
+//    rkey = htobe32(ibmr->rkey);
```
Removed

foraxe committed:
Fix EOF in gpunetio_kernels.cu

brminich commented on src/plugins/gpunetio/gpunetio_backend_aux.h:
```diff
@@ -87,6 +87,7 @@ struct docaXferReqGpu {
     uint32_t rkey[DOCA_XFER_REQ_SIZE];
     uint16_t num;
     uint8_t in_use;
+        uint8_t keys_are_swapped;
```
alignment

foraxe committed:
Address new GPUNetIO review comments

- Drop NIXL_GPUNETIO_DEBUG_DUMP/debug_dump gating; use NIXL_DEBUG/NIXL_TRACE directly
- Simplify hex logging (avoid format_hex helper) and fix alignment issues in gpunetio_backend_aux.h
- Update SPDX copyright year to 2026 in modified GPUNetIO files

foraxe committed:
Apply clang-format fixes

foraxe committed:
Merge upstream/main

foraxe committed:
SPDX: update copyright year to 2026
